configuration.md
---
layout: default
title: Circleator - Configuration File Reference
---

# Circleator - Configuration File Reference

## Introduction

The Circleator configuration file, specified by the `--config` command line
option, determines almost every aspect of how Circleator will draw the figure.
This document describes the configuration file format in detail, and covers 
the following topics:

* [File Format](#file-format)
* [Predefined Tracks](#predefined-tracks)
* [User Defined Tracks](#user-defined-tracks)
* [Track Size and Position](#track-size-and-position)
* [Feature Selection](#feature-selection)
* [Colors](#colors)
* [Labels](#labels)
* [Opacity](#opacity)
* [Loops](#loops)
* [Z-Index](#zindex)

### File Format

The Circleator configuration file is a tab-delimited plain
text file designed to be edited manually. Blank lines or lines
beginning with the "#" character (similar to Perl-style comments) are
ignored. Every other line must contain at least one of the following
tab-delimited fields (almost all of which are optional):

1. **type** - either the name of a [predefined track type][predef_tracks] OR the keyword `new`
2. **name** - a name by which the track may be referenced from elsewhere in the configuration file
3. **glyph** - the Circleator "glyph" used to render this track
4. **heightf** - the height of the track as a fraction of the circle's radius (0-1)
5. **innerf** - position of the innermost part of the track as a fraction of the circle's radius (0-1)
6. **outerf** - position of the outermost part of the track as a fraction of the circle's radius (0-1)
7. **data** - path to a data file, if one is required by the chosen track **type** and/or **glyph**
8. **feat_type** - display only features of the specified type (e.g., "gene", "tRNA")
9. **feat_strand** - display only features on the specified strand (e.g., "-", "+", "-1", "1")
10. **color1** - interpretation depends on the track type: usually the SVG fill color
11. **color2** - interpretation depends on the track type: usually the SVG stroke (outline) color
12. **opacity** - opacity of the track between 0 and 1, where 0 = invisible/completely transparent and 1=completely opaque
13. **zindex** - integer z-index of the track: tracks with higher z-indexes are drawn on top of those with lower z-indexes
14. **options** - a comma-delimited list of track options in the format "key=value"

[predef_tracks]: {{site.baseurl}}/predefined-tracks.html

All of these fields except for the very first one (**type**) are
optional and may be omitted completely. Hence the simplest possible
line/track in a Circleator configuration file is the name of a
predefined track type and nothing else, like this:

    coords

You can choose to include as many or as few of the field values as you
like, provided that the **type** is always specified. But note that
once you start skipping fields, you have to skip all the rest too. So,
for example, you could choose to include only fields 1-3 (**type**, 
**name**, **glyph**)

    new	r1	rectangle

Or you could specify only fields 1-6 (**type**, **name**, **glyph**,
**heightf**, **innerf**, **outerf**):

    new	r1	rectangle	null	0.7	0.75

But you *cannot* specify only fields 1-3 and 10 (**type**, **name**,
**glyph**, **color1**) without including all the blank fields between
numbers 3 and 10:

<pre class='illegal'>
new	r1	rectangle	#ff0000
</pre>

The one very important exception to this rule is that you are allowed
to include the final **options** field, even if you've skipped some of
the fields that precede it. For example, you can specify fields 1-3
plus the final **options** field (#14):

    new	r1	rectangle	stroke-width=3

Another useful feature is that any of the positional fields listed
above, with the exception of the first and last, can be included (by
name) in the final **options** field. So while you *can't* specify
values for fields 1-3 and 10 like this (*i.e.*, with only a single tab
between `rectangle` and the color specifier `#ff0000`):

<pre class='illegal'>
new	r1	rectangle	#ff0000
</pre>

You *can* do it like this:

    new	r1	rectangle	color1=#ff0000

Or, equivalently, like this (note the lack of spaces between the
different options in the comma-separated **options** list: the
configuration file parser does not allow spaces to be used here):

    new	name=r1,glyph=rectangle,color1=#ff0000

Note that in one of the preceding examples we used the word `null` to
indicate a missing field value. Here are some other ways to indicate
missing/absent field values:

* Leaving the field--and all those that follow it--out of the line and skipping directly to the **options** field.
* Using one of the following equivalent ways to indicate a null/undefined value: `null`, `undef`, `n/a`, `na`, `n`, `.` (period), `-` (single dash)

### Predefined Tracks
(relevant options: **type**)

If the value in the first column of a line in the configuration file
is anything other than `new` then it is the name of a predefined track
type.  The entire set of Circleator-supported predefined track types
is listed on the [predefined tracks page][predef_tracks], along with
information on the most commonly-used configuration options for each
track type.

[predef_tracks]: {{site.baseurl}}/predefined-tracks.html

### User Defined Tracks
(relevant options: **type**, **name**, **glyph**)

If the value in the first column of a line in the configuration file
is `new` then the track is a user-defined track and must include
columns 2 (**name**) and 3 (**glyph**), either as positional fields or
named values in the **options** field.  Here are the
currently-supported Circleator glyphs:

 * ruler
 * rectangle
 * label
 * graph
 * scaled-segment-list
 * load
 * load-bsr
 * bsr
 * load-trf-variation
 * load-gene-expression-table
 * load-gene-cluster-table
 * compute-deserts
 * compute-graph-regions
 * cufflinks-transcript
 * synteny-arrow
 * loop-start
 * loop-end

The prdefined track types are all defined in terms of these glyphs, so
a predefined track type is really nothing more than an alias for a
user-defined track (plus possibly some default values for some of the
track options). The **coords** track, for example, is defined like
this in the system-wide predefined track configuration file
([conf/predefined-tracks.cfg][predef_tracks_file]):

    new coords ruler 0.02 tick-interval=100000,label-interval=500000,label-type=curved

This means that placing either of the following two lines into a
configuration file will have exactly the same effect:

    # use predefined track type alias for the 'ruler' glyph:
    coords
    # create user-defined 'ruler' track with the exact same options:
    new c1 ruler 0.02 tick-interval=100000,label-interval=500000,label-type=curved

Something to note here is that the system-wide predefined track
configuration file is not the only place where new track types can be
defined: every time you create a user-defined track and give it a
unique name (e.g., `c1` in the above example) that name can then be
reused in the same configuration file (strictly *after* the line on
which it first appears) as though it were a predefined track
type. This can be useful in cases where you wish to define a number of
very similar tracks but without having to retype (or cut and paste)
all the track options.  For example, in this configuration file
excerpt one of the user-defined tracks is given the name
`lightblue_ring` and then that name is used later in the same file to
produce another copy of the same track:

    genes
    small-cgap
    new lightblue_ring rectangle feat-type=contig,color1=blue,opacity=0.2,heightf=0.05
    small-cgap
    # "lightblue_ring" is now an alias for the track defined above
    lightblue_ring

Currently the best way to begin learning about the underlying glyphs
is to examine the file that *defines* the predefined track types using
the glyphs. This file, called
[predefined-tracks.cfg][predef_tracks_file], is itself a Circleator
configuration file, albeit with a significant amount of documentation
and special processing directives embedded in its comments.

[predef_tracks_file]: https://github.com/jonathancrabtree/Circleator/blob/master/conf/predefined-tracks.cfg

### Track Size and Position 
(relevant options: **heightf**, **innerf**, **outerf**)

There are 3 configuration file options that control *where* a track will appear.  Each track
is defined as the area between two concentric circles, where the distance between those circles and
their individual radii are specified by **heightf**, **innerf**, and
**outerf**, respectively. These three options are not independent: given
any two of them the third can be calculated according to the following very simple equation:

    heightf = outerf - innerf

In practical terms what this means is that if you specify all three of
them in the configuration file then Circleator will ignore one of
them. If you choose not to supply *any* of these options then the
Circleator will assign a default **heightf** value to the track, which
determines how thick the track will be, and it will set the **outerf**
value so that the track appears just *inside* the track that preceded
it in the configuration file.  So if the configuration file contains a
number of tracks with no positioning or size information then by
default the first track in the file will appear at the very edge of
the circle (**outerf** = 1.0) and successive tracks will be nested
inside those that came first. If too many tracks are placed using this
method it is possible to run out of space in the circle (*i.e.*, for a 
track to have a computed **outerf** value < 0), in which case
Circleator will report an error when it tries to draw the figure.

The following figure shows a number of tracks that are labeled with
their **innerf** - **outerf** values. It also illustrates how one
track can be drawn on top of another by using transparency and an
**innerf** - **outerf** range that overlaps with one or more other
tracks:

![Track size illustration][CM000961_tracks_med]  
Here is the configuration file for the above figure: [tracks-1.cfg][CM000961_tracks_conf]  

[CM000961_tracks_med]: {{site.baseurl}}/images/CM000961-tracks-1-800.png
[CM000961_tracks_conf]: {{site.baseurl}}/tracks-1.cfg

Going roughly from outermost to innermost the tracks in this figure are:

* the 3 thin grey lines indicating gaps are all in a track from 0-1.1 i.e., **innerf**=0, **outerf**=1.1, **heightf**=1.1: this track overlaps with all of the other tracks
* outermost blue "contigs" track from 0.9-1.0 i.e., **innerf**=0.9, **outerf**=1.0,  **heightf**=0.1
* middle blue "contigs" track from 0.7-0.8 i.e., **innerf**=0.7, **outerf**=0.8,  **heightf**=0.1
* thin blue "contigs" track from 0.55-0.6 (i.e., **innerf**=0.55, **outerf**=0.6,  **heightf**=0.05
* thin grey track from 0.5-0.55 i.e., **innerf**=0.5, **outerf**=0.55,  **heightf**=0.05
* green transparent track from 0.5-1.0 i.e., **innerf**=0.5, **outerf**=1.0,  **heightf**=0.5: this track overlaps with all of the above tracks
* inner blue circle from 0-0.4 i.e., **innerf**=0, **outerf**=0.4,  **heightf**=0.4

In this example one of the tracks (the `gap` track) has an **outerf**
value of 1.1. There's nothing wrong with using values greater than 1,
but at some point these tracks will run off the edge of the
figure. How soon that happens is a function of the `--pad`
command-line argument, which determines how much blank space the
Circleator will leave on each side of the circle. By default the
circle has a diameter of 2400 and a `--pad` value of 400. Note also
that any of these three track configuration options (**innerf**,
**outerf**, and **heightf**) can be set to the special value **same**,
which means that they will take on the same value as the track that
immediately *preceded* the current track in the configuration
file. Setting both **innerf** and **heightf** to **same**, for
example, will create a track that *exactly* overlaps the one before
it. One can also add or subtract a constant value from **same** to
specify a value that is relative to the previous track. For example,
setting `innerf=same+0.1` specifies an inner fraction that is 0.1
larger than the inner fraction of the previous track.

### Feature Selection 
(relevant options: **feat_type**,**feat_strand**, and others)

Many track and glyph types, for example the commonly-used `rectangle`
glyph, operate on sets of features. For example, the `genes` track is
defined thusly:

    new genes rectangle 0.07 . . . gene . #000000

Note the value of `gene` in the **feat_type** field and "." in the
**feat_strand** field; this indicates that a black (`#000000`) curved
rectangle (the `rectangle` glyph) should be drawn for every input
feature with type "gene", regardless of strand.  The definition for
the `genes-fwd` track is the same but with **feat_strand** set to "1"
to draw only forward-strand genes:

    new genes-fwd rectangle 0.07 . . . gene 1 #000000

As usual, the **feat_type** and **feat_strand** may be included in the
final **options** field rather than using the equivalent positional
options. Note that either dashes or underscores may be used in the
final options field: **feat-type** and **feat_type** are treated the same:

    new genes-fwd-too rectangle 0.07 feat-type=gene,feat-strand=1,color1=#000000

#### Feature Filter Track Options

**feat_type** and **feat_strand** are both examples of feature
"filters". That is, they filter or restrict the set of features on
which the track will operate. The following list enumerates the other
feature filters that Circleator supports:

* **refseq-name** - Filters out all features that are NOT from the specified sequence/contig  

       # draw only genes on contig239
       new genes rectangle heightf=0.05,feat-type=gene,refseq-name=contig239

* **clip-fmin**, **clip-fmax** - Filters out all features that are outside the specified sequence coordinate range  

        # label all genes in the region between 265kb and 275kb with their locus_ids:
        small-label label-type=spoke,innerf=1.04,feat-type=gene,label-function=locus,clip-fmin=265000,clip-fmax=275000

* **feat-type** - Filters out all features that are NOT of the specified type (e.g., exon, CDS, gene, rRNA, tRNA)  

        # color all tRNAs red
        new genes rectangle heightf=0.05,color1=red,feat-type=tRNA

* **feat-type-regex** - Filters out all features whose type is NOT matched by the specified Perl regular expression  

        # color all tRNAs AND all rRNAs red
        new genes rectangle heightf=0.05,color1=red,feat-type-regex=[tr]RNA

* **feat-strand** - Filters out all features that are NOT on the specified strand  

        # color all forward-strand tRNAs red
        new genes rectangle heightf=0.05,color1=red,feat-type=tRNA,feat-strand=1

* **feat-min-length**, **feat-max-length** - Filters out all features whose length (i.e., max genomic coord - min genomic coord) is less than (or more than) the specified value

        # label all genes that are more than 20kb long:
        small-label label-type=spoke,innerf=1.04,feat-type=gene,label-function=locus,feat-min-length=20000

* **overlapping-feat-type** - Filters out all feature that do NOT overlap with another feature of the specified type

       # label all CDS features that contain at least one SNP
       new GL label feat-type=CDS,innerf=1.32,heightf=0.02,label-function=product,label-type=spoke,packer=none,overlapping-feat-type=SNP

* **feat-tag** - Used in conjunction with the remaining filters to select features based on their attributes. If 
none of the other feat-tag filters are given then it will filter all features that do NOT have the specified attribute.

* **feat-tag-value** - Filter out all features whose **feat-tag** attribute does NOT have the specified value

        # draw only SNPs with SNP_num_diffs=7 (i.e., SNPs at which 7 of the query genomes differ from the reference)
        new SNP rectangle 0.03 feat-type=SNP,stroke-width=1,color1=#000000,feat-tag=SNP_num_diffs,feat-tag-value=7

* **feat-tag-min-value**, **feat-tag-max-value** - Filter out all features whose **feat-tag** attribute is less than (or more than) the specified value

        # highlight genes whose expression value in the lung-derived sample is greater than 2
        new H1 rectangle innerf=0.32,outerf=1.13,opacity=0.5,color1=#d0d0d0,color2=#000000,feat-type=CDS,feat-tag=EXP_Lung,feat-tag-min-value=2

* **feat-tag-regex** - Filter all features whose **feat-tag** attribute does NOT match the specified Perl regular expression

        # highlight genes whose gene product matches either "hypothetical protein" or "hypotheticalprotein"
        new HCH rectangle feat-type=CDS,feat-tag=product,feat-tag-regex=hypothetical\sprotein,innerf=same,outerf=same,opacity=0.8,color1=#00ff00

#### Feature Source

In all of the preceding examples we are implicitly considering *all*
of the genomic sequence features that appear in the original input
files (i.e., the file(s) passed to the `--data` and/or `--contig_list`
command line options) and selecting a subset of those features based
on feature type, feature strand, feature attribute values, etc. If the
original input file is a GenBank flat file, for example, then features
of the following types are common (this is not meant to be an
exhaustive list): **source**, **gap**, **gene**, **CDS**, **tRNA**,
**rRNA**, **ncRNA**, **tmRNA**, **misc_feature**, **misc_binding**. In
addition to features that appear explicitly in the input file(s),
Circleator also adds some features to represent the genomic
sequences/contigs and the spaces between them:

* **reference_sequence** - Circleator will always create a single feature of type **reference_sequence** that spans the *entire* reference coordinate system, including gaps if there are multiple sequences.
* **contig** - Circleator will create a feature of type 'contig' for each input sequence specified by the `--data`, `--sequence`, or `--contig_list` options.
* **contig_gap** - If the input contains multiple contigs then Circleator will create a feature of type 'contig_gap' between each pair of adjacent contigs (even if `--contig_gap_size` is set to 0). If the `--contig_list` option is used then the location and size of individual gaps can be specified explicitly in the contig list file.
* **genome** - Circleator will add a feature of type 'genome' for each 'genome' line in the contig list file specified by `--contig_list`. See the [command line documentation][cmdline] for more information on the use of this option.

[cmdline]: {{site.baseurl}}/command-line.html

With this in mind, here are the possible feature sources for any given track:

1. *The combination of all the features from all of the input files plus the special features mentioned above.*  
This is the default feature source.

2. *The set of features that are used by another track in the figure.*  
This feature source can be selected with the **feat-track** option, which specifies the (hopefully unique) **name** of the track whose feature list should be 
used. A common use of this option is to define a track that prints a label for every feature that was drawn in a previous track. Using the
**feat-track** option is essentially a shortcut to avoid having to repeat all of the same feature filters in multiple tracks.  Here is a simple example:

        # we use the predefined track type 'contigs' to show all features of type contig 
        # (i.e., feat-type=contig) and name it "c1"
        contigs c1
        # label only the contig features that were selected by track c1:
        medium-label feat-track=c1,label-function=primary_id,packer=none,innerf=same+0.01,outerf=same,text-color=#ffffff

3. *The set of features loaded from a (new) external file.*  
The **feat-file** and **feat-file-type** options allow one to load
features from files that were not included in the `--data`,
`--sequence`, or `--contig_list` command-line options that were given
when Circleator was run.  The **feat-file** option specifies the full
path to the file to be loaded and the **feat-file-type** option
specifies what type the file is. If the file is a format accepted by
BioPerl (e.g. GenBank flat file, multi-FASTA) then **feat-file-type**
may be omitted. Otherwise it must be one of the following
(case-insensitive) options:
   * ucsc_refGene
   * ucsc_refGene_exons
   * ucsc_knownGene
   * ucsc_knownGene_exons
   * ucsc_rmsk
   * cufflinks_gtf
   * skirret-snp
   * merged-table-snp
   * snp-table
   * VCF
   * csv-snp
   * tabbed-snp
   * trf
   * gff

   For example:

        # Load SNPs for ATCC_30222 from VCF:
        new r1 rectangle 0.06 feat-file=data/ATCC_30222.extra-filtered.vcf,feat-file-type=VCF,feat-type=SNP,color1=snp_type,color2=snp_type,snp-query=ATCC_30222

4. *A single user-defined feature.*  
It is also possible to define a single simple feature directly in the configuration file by using
the following track options:
   * **user-feat-fmin**, **user-feat-fmax** - coordinates of the feature in chado-style 0-indexed interbase coordinates
   * **user-feat-start**, **user-feat-end** - coordinates of the feature in BioPerl-style 1-indexed base-based coordinates
   * **user-feat-strand** - strand of the user-defined feature, either 0,1, or -1
   * **user-feat-type** - type of the user-defined feature (e.g., "gene", "new_feat_type_x")
   * **user-feat-seq** - reference sequence/contig on which the feature is localized
   * **user-feat-id** - unique id for the user-defined feature

    This can be useful for defining regions of interest directly in the configuration file.  For example, the following configuration
    file excerpt highlights the user-defined sequence region from 0-5kb:

        new SD5000 rectangle innerf=0,outerf=1.2,opacity=0.15,color1=#ff00ff,user-feat-fmin=0,user-feat-fmax=5000

    One could also define several user-defined features (at most one per line) with the same **user-feat-type**
    and then refer to all of those features in a subsequent track using **feat-type**. For example, here we 
    create several ROI (region of interest) features and then highlight them. Note the use of the **load**
    glyph/track type, which does not draw anything, but only loads and/or creates new features:

        new uf1a load user-feat-fmin=3898103,user-feat-fmax=3898620,user-feat-type=roi
        new uf2a load user-feat-fmin=844688,user-feat-fmax=845220,user-feat-type=roi
        new highlight_rois rectangle feat-type=roi,innerf=0,outerf=1.1,opacity=0.4,color1=red,color2=black

**NOTE:** Certain track types can be thought of as having both "input"
and "output" features. That is, they select a subset of the available
features by using feature filters (the input) and then they produce a
new set of features (the output) by processing the input features in
some way. For example, the **compute-deserts** track type takes a set
of input features and then creates a new "desert" feature wherever it
finds a large region that is empty of the input features. It could be
used to create a feature of type "SNP_desert" wherever there is 15kb
or more of sequence that does not contain any SNPs: the SNPs are the
input features and the SNP_deserts are the output features. Currently
the **feat-track** option always refers to the input features for a
track, never the output features. In the preceding example one could
instead refer to the output desert features by using the feature
filter `feat-type=SNP_desert`. Also in some cases the **feat-type**
option is used to specify the type of features that a track should
create, rather than filtering the input features.

### Colors
(relevant options: **color1**, **color2**, **text-color**)

Wherever it expects a color specification (e.g., in the **color1**, **color2**, and **text-color** options) 
Circleator will accept any of the following:

1. *Any valid [SVG 1.1 color specifier][svg_colors]*  
For example, all of the following are valid ways to set **color1** to red:

        new cr1a rectangle feat-type=gene,color1=rgb(255,0,0)
        new cr1b rectangle feat-type=gene,color1=red
        new cr1c rectangle feat-type=gene,color1=#ff0000

2. *Any color function defined in Circleator's Circleator::FeatFunction::Color package:*  


ChangeLog

1.0.0rc5 (Apr ?? 2014)
 -Add parser for clc_find_variations SNP/indel-finding tool.
 -Add feat-tag-gt, feat-tag-gte, feat-tag-lt, feat-tag-lte. Previously only 
  lte (less than or equal to) and gte (greater than or equal to) were supported, via
  feat-tag-min-value and feat-tag-max-value.
 -Add overlapping-feat-track option.
 -Add variant-base-histogram glyph

1.0.0rc4 (Oct 15 2013)
 -Add --version option to report the current version/release number.
 -Modify Build.PL so it doesn't install circleator.bak into /usr/local/bin/.

1.0.0rc3 (Oct 11 2013)
 -Add missing MANIFEST file to avoid build error.

1.0.0rc2 (Oct 11 2013)
 -Fix Build.PL to work with system-wide installs [Issue #12]
 -Add --conf_dir option to circleator
 -Modify circleator.t to save temporary/log dir if test(s) fail [Issue #2] 

1.0.0rc1 (Oct 02 2013)
 -Changed release numbering syntax to the more standard major.minor.revision
 -Moved code to GitHub (https://github.com/jonathancrabtree/Circleator)
 -Moved ExtJS GUI into separate GitHub repository ("Ringmaster")
 -Added Build.PL
 -Reinstated unknown sequence id warning when parsing flat file graph/heat map data.
 -Fixed bug in get_track_features causing feat-track references to be incorrectly processed.
 -Added font-width-frac option to allow overriding FONT_WIDTH_FRAC global on a per-track basis.

circleator-1.0alpha11 (Jun 04 2013)
 -Bug fix: label-units=Gb was using the label "Bb" instead of "Gb"
 -label-units is now case-insensitive i.e., label-units=mb and label-units=Mb will both work
 -Added new tests for label-units

circleator-1.0alpha10 (Jun 03 2013)
 -Added tests for: graph-type=line, graph-type=heat_map
 -Bug fix for case where graph-type=heat_map and no colors or color palette are supplied
 -Fixed typos in references to GCskew tracks in the predefined track documentation.
 -Corrected track options (heat_map_*, target_bp) to use '-' instead of '_' (but code now allows either)
 -Bug fix: ignore graph_baseline when graph-type=heat_map
 -Bug fix: fixed discrete color assignment for graph-type=heat_map
 -Added "Gb" as new option for label-units.
 -Bug fix: user-feat-type option was being ignored.
 -Added 'tag' label function for displaying arbitrary feature attributes/tags e.g., GFF attributes such as 'ID'.

circleator-1.0alpha9 (May 15 2013)
 -Added 'heat_map' graph type plus support for Brewer color palettes from http://www.colorbrewer.org

circleator-1.0alpha8 (Apr 26 2013)
 -Modified SNP_Table to allow IUPAC ambiguity codes.

circleator-1.0alpha7 (Apr 26 2013)
 -Added compute-graph-regions glyph: identifies regions where graph value is above/below specified bounds.
 -Added overlapping-feat-type filter: only features that overlap with a feature of the specified type will be selected.
 -Modified admin/makePredefTrackDocs.pl to regenerate SVG and PNG images only for updated config files
 -Updated predefined track documentation: added label-units, graph-type
 -Improved scaled_segment_list warning when scale is set too high.
 -Circleator now automatically reduces scaled-up regions to 75% of genome whenever they would otherwise take >= 100%
 -Added parser for new gene cluster file format (Gene_Cluster_Table.pm), new load-gene-cluster-table glyph
 -Added new options to support filtering genes by gene cluster signature:
  gene-cluster-genomes, gene-cluster-signature, gene-cluster-analysis, gene-cluster-min-genomes, gene-cluster-max-genomes
 -Added parser (SNP_Table.pm) for new SNP file format, "snp-table", compatible with CloVR-Comparative SNP output.

circleator-1.0alpha6 (Mar 20 2013)
 -Bug fix for issue with parameters not getting passed to Circleator::SeqFunction classes.
 -Added missing feature filters (clip-fmin, clip-fmax) to Circleator/Config/Standard.pm
 -Bug fix for loading features onto reverse-complemented contigs: they are now reverse complemented too.
 -Changed log message to not treat 'gap' and 'genome' features in --contig_list as contigs.
 -Added line graph support to graph glyph.
 -Added 'snp-no-hit-label' and 'snp-same-as-ref-label' options to snp_base.pm
 -Updated snp_base and snp_base_no_indel SNP color functions to handle multiple alleles.
 -Updated MergedTable SNP parser to explicitly record "No Hit" values.
 -Updated MergedTable SNP parser to handle multiple alleles separated by "," or "/".
 -Added computed 'num_no_hits' attribute to parsed SNPs.

circleator-1.0alpha5 (Feb 20 2013)
 -primary_id is now set correctly in Circleator-created features (e.g., 'contig', 'genome'.)
 -Added missing feature filters (feat-min-length, feat-max-length) to Circleator/Config/Standard.pm
 -Fixed bug in Circleator/Util/Deserts.pm: compute-deserts glyph was throwing an error on linear sequences.
 -Bug fix and changes to accomodate a new variant of the Skirret format.
 -Added conf/clovr-microbe and 3 configuration files for the CloVR Microbe pipeline's output.

circleator-1.0alpha4 (Feb 15 2013)
 -Minor documentation improvements.
 -Added snp-intronic-color, snp-readthrough-color options to the snp_type color function.
 -Added snp_type_no_indel color function, which is the same as snp_type but without the insertion and deletion colors.
 -Changed the feature type of contig gaps (in multi-contig figures) from 'contig' to 'contig_gap'.
 -Fixed off-by-one bug in the positioning of 'contig' and 'contig_gap' features in multi-contig figures.
 -Added 'gff' feat-file-type, which uses the BioPerl GFF parser in Bio::FeatureIO::gff.
 -Added contig-gaps predefined track.
 -Added data/AE003852-AE003853.gbk, chrI and chrII of V. cholerae O1 biovar eltor str. N16961
 -Fixed bug whereby additional feature filters on tracks using 'feat-track' were ignored.
 -Added the following categories to the --debug command-line option: coordinates,input,loops,misc,packing
 -Added optional 'revcomp' flag to --contig_list file format: including it on a line reverse complements the corresponding input contig.

lib/Circleator/Config/Config.pm
package Circleator::Config::Config;

my $DEFAULT_MAX_END_FRAC = 1;

# TODO - add method and/or option to rescale track starts and/or ends into a specified range (e.g., 0-1)

# ensure each track has a unique name
sub update_track_names {
  my($tracks, $logger) = @_;
  my $tracks_by_name = {};

  my $nt = scalar(@$tracks);
  for (my $t = 0;$t < $nt;++$t) {
    my $track = $tracks->[$t];
    my $tname = $track->{'name'};
    # record original, possibly non-unique name
    $track->{'original_name'} = $tname;
    # assign nameless tracks a number
    if (!defined($tname) || ($tname =~ /^\s*$/)) {
      $track->{'name'} = $t+1;
    }
    else {
      my $list = $tracks_by_name->{$name};
      $list = $tracks_by_name->{$name} = [] if (!defined($list));
      push(@$list, $track);
    }
  }  

  # append a .1, .2., .3 to any duplicates
  foreach my $name (keys %$tracks_by_name) {
    my $list = $tracks_by_name->{$name};
    my $ll = scalar(@$list);
    if ($ll > 1) {
      my $suffix = 1;
      map { $_->{'name'} .= "." . $suffix++; } @$list;
    }
  }
}

# compute explicit start and end fraction values for each track
sub update_track_start_and_end_fracs {
  my($tracks, $logger, $debug) = @_;

  # compute start-frac and end-frac for those not explicitly specified
  my $nt = scalar(@$tracks);
  my $last_start_frac = $DEFAULT_MAX_END_FRAC;
  my $last_end_frac = $DEFAULT_MAX_END_FRAC;

  for (my $t = 0;$t < $nt;++$t) {
    my $track = $tracks->[$t];
    my $tname = $track->{'name'};
    my $tglyph = $track->{'glyph'};
    # save original values in case positions have to be recomputed (e.g., for loop unrolling - see Circleator::Util::Loop)
    my($sf, $ef, $hf) = map { $track->{'original_' . $_} = $track->{$_}; } ('start-frac', 'end-frac', 'height-frac');
    my($sfs,$efs,$hfs) = map { defined($_) ? $_ : "" } ($sf, $ef, $hf);
    $logger->logdie("height-frac cannot be negative (set to $hfs for track ${tname})") if (defined($hf) && ($hf < 0));
    $logger->debug("track $tname before update: sf=$sfs ef=$efs hf=$hfs glyph=$tglyph") if ($debug);

    # TODO - turn this back into a warning _except_ for predefined tracks?
    if (defined($sf) && defined($ef) && defined($hf)) {
      $logger->debug("innerf ($sf), outerf ($ef), and heightf ($hf) are all specified for track ${tname}, ignoring height-frac") if ($debug);
      $hf = undef;
    }

    # if either sf or ef is 'same' it gets set to the same value as the previous track
    # sf or ef can also be set to some offset from the previous track's value e.g., "same+0.2" or "same-0.1"
    my $eval_frac = sub {
      my($frac, $last_frac) = @_;
      if ($frac eq 'same') {
        return $last_frac;
      } elsif ($frac =~ /^same\s*([+\-])\s*([\d\.]+)\s*$/) {
        my($op, $num) = ($1, $2);
        return ($op eq '+') ? $last_frac + $num : $last_frac - $num;
      } else {
        return $frac;
      }
    };

    if (defined($sf) && ($sf =~ /^same/)) {
      $sf = $track->{'start-frac'} = &$eval_frac($sf, $last_start_frac);
    }
    if (defined($ef) && ($ef =~ /^same/)) {
      $ef = $track->{'end-frac'} = &$eval_frac($ef, $last_end_frac);
    }

    # if $ef and $sf are omitted but $hf is not then $ef is set to the last start_frac
    if ((!defined($ef)) && (!defined($sf))) {
	# default height of 0 for tracks that don't actually display anything
	if (!defined($hf) && ($tglyph =~ /^(loop-|compute|load|scaled-segment-list)/)) {
	    $hf = 0;
	}
	if (defined($hf)) {
	    $ef = $track->{'end-frac'} = $last_start_frac;
	}
    }

    # if $hf is set it can be used in place of one of the endpoints
    if (defined($hf)) {
      if (!defined($sf)) {
        $sf = $track->{'start-frac'} = $ef - $hf;
      } elsif (!defined($ef)) {
        $ef = $track->{'end-frac'} = $sf + $hf;
      }
    }

    $logger->logdie("either heightf or innerf and/or outerf must be defined for track ${tname} with glyph=$tglyph") if (!defined($sf) || !defined($ef));
    $logger->logdie("innerf ($sf) > outerf ($ef) for track ${tname}") if ($sf > $ef);
    $logger->logdie("innerf ($sf) < 0 for track ${tname}") if ($sf < 0);
    # TODO - glyphs should be full-fledged objects, one of whose properties indicates whether they're typically 0-height
    $logger->warn("innerf ($sf) = outerf ($ef) for track ${tname} with glyph=$tglyph") if (($sf == $ef) && ($tglyph !~ /^(loop-|compute|load|scaled-segment-list|none)/));
    $logger->debug("Config track $t/$tname sf=$sf ef=$ef") if ($debug);

    $last_start_frac = $sf;
    $last_end_frac = $ef;
  }
}

sub index_tracks {
  my($tracks, $logger, $key) = @_;
  my $index = {};
  my $nt = scalar(@$tracks);
  for (my $t = 0;$t < $nt;++$t) {
    my $track = $tracks->[$t];
    my $tname = $track->{$key};
    my $list = $index->{$tname};
    $list = $index->{$tname} = [] if (!defined($list));
    push(@$list, $track);
  }
  return $index;
}

# $fn_factories - hashref mapping track attribute prefix (e.g., "color", "label-function") to a 
# coderef that takes as input a function name and returns a factory coderef that maps from 
# ($track, $att_name) to the function/coderef of the appropriate type.
#
sub update_track_function_refs {
  my($tracks, $logger, $fn_factories) = @_;

  # replace function keywords with actual (parameterized) functions
  foreach my $track (@$tracks) {
    foreach my $att (keys %$track) {
      my $tval = $track->{$att};
      foreach my $ff_re (keys %$fn_factories) {
        if ($att =~ /$ff_re/) {
          my $ff_fn = $fn_factories->{$ff_re};
          my $ff = &$ff_fn($tval);
          if (defined($ff)) {
            # replace function name with the actual function
            $track->{'original_fn_' . $att} = $tval;
            $track->{$att} = &$ff($track, $att);
            last;
          }
        }
      }
    }
  }
}

1;
